PLFM-8986

nickgros opened the pull request:
Bootstrap the JSON schemas for all entity types.

Note: base branch is release-542 (production)

nickgros committed:
PLFM-8986

Bootstrap the JSON schemas for all entity types.

marcomarasca reviewed (approved):
LGTM

nickgros committed:
PLFM-8986

Fix loadAllSchemasRecursive to walk through objects that do not have `$ref`/ `$id`

nickgros commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/SynapseSchemaBootstrapImpl.java:
```diff
@@ -181,36 +202,59 @@ Optional<Long> getNextPatchNumberIfNeeded(String organizationName, String schema
-		if (!visitedId.add(schemaClassName)) {
-			return;
+
+		// Walk through the schema to continue loading all references
+		schema.getSubSchemaIterator().forEachRemaining((ObjectSchema subSchema) ->
+				loadAllSchemasRecursive(loadedSchemas, subSchema.getRef(), visitedId, subSchema)
+		);
```
Previous implementation did not walk through inlined subschemas (e.g. subschemas that were not a simple `$ref`). Fix this by also passing the subschema in the recursive call.

For example, in the below schema, the previous implementation did not visit `org.sagebionetworks.repo.model.MyItemType` because the schema for `$.properties.myArrayProperty` is inlined instead of referenced.

```json
{
  "$id": "org.sagebionetworks.repo.model.MyObject",
  "properties": {
    "myArrayProperty": {
      "type": "array"
      "items": {
        "$ref": "org.sagebionetworks.repo.model.MyItemType"
      }
    }
  }
}
```

marcomarasca commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/SynapseSchemaBootstrapImpl.java:
```diff
@@ -181,36 +202,59 @@ Optional<Long> getNextPatchNumberIfNeeded(String organizationName, String schema
+	 * @param visitedId the set of visited schema classes to prevent infinite loops
+	 * @param schema the schema to walk through (ignored if schemaClassName is provided)
+	 */
+	private void loadAllSchemasRecursive(Map<String, ObjectSchema> loadedSchemas, String schemaClassName,
```
Maybe we can get rid of this "alternative" schemaClassName and just use the id of the "ObjectSchema schema" to simplify this?

nickgros committed:
PLFM-8986 - code review changes

Simplify `loadAllSchemasRecursive` to only take a schema, rather than an ID or schema.
